From: Modestas Vainius <modestas@vainius.eu>
Subject: [PATCH] Link MySQLe (libmysqld) to amarok executable.

When libmysqld.a is linked to amarok executable, it no longer needs
to be built with -fPIC. This avoid portability problems at low cost.
sqlcollection is essential Amarok plugin anyway.

libmysqld.a as of MySQL 5.1.34 contains duplicate symbols (from 
ha_federated.o and libfederated_a-ha_federated.o). This causes
trouble linking on alpha. Hence we also need to locally copy and 
remove ha_federated.o from libmysqld.a before linking.

Signed-off-by: Modestas Vainius <modestas@vainius.eu>

---
 src/CMakeLists.txt                          |   20 +++++++++++++++++++-
 src/collection/sqlcollection/CMakeLists.txt |   18 +++---------------
 2 files changed, 22 insertions(+), 16 deletions(-)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index a4788e7..ed179ad 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -599,8 +599,26 @@ if(Q_WS_MAC)
     SET_SOURCE_FILES_PROPERTIES(${CMAKE_CURRENT_BINARY_DIR}/amarok.icns PROPERTIES MACOSX_PACKAGE_LOCATION Resources)
     install(TARGETS Amarok ${INSTALL_TARGETS_DEFAULT_ARGS})
 else(Q_WS_MAC)
-    kde4_add_executable(amarok ${amarok_SRCS})
+    # libmysqld has duplicated symbols as of MySQL 5.1.34.
+    # Copy the lib and remove dupes (they come from ha_federated.o).
+    set(_local_libmysqld "${CMAKE_CURRENT_BINARY_DIR}/libmysqld.a")
+    add_custom_command(OUTPUT ${_local_libmysqld}
+                       COMMAND ${CMAKE_COMMAND} -E copy ${MYSQL_EMBEDDED_LIBRARIES} ${_local_libmysqld}
+                       COMMAND ar dv ${_local_libmysqld} ha_federated.o
+                       COMMAND ranlib ${_local_libmysqld}
+                       DEPENDS ${MYSQL_EMBEDDED_LIBRARIES}
+                       COMMENT "Locally copying and fixing libmysqld.a")
+
+    kde4_add_executable(amarok ${amarok_SRCS} ${_local_libmysqld})
     target_link_libraries(amarok ${KDE4_KDEUI_LIBS} amaroklib )
+    # Link MySQLe to the executable. Hence -fPIC for libmysqld is not needed.
+    target_link_libraries(amarok -Wl,-whole-archive ${_local_libmysqld} -Wl,-no-whole-archive)
+    if(NOT WIN32)
+        if(NOT APPLE)
+            target_link_libraries(amarok crypt pthread dl wrap)
+        endif(NOT APPLE)
+    endif(NOT WIN32)
+
     install(TARGETS amarok ${INSTALL_TARGETS_DEFAULT_ARGS})
 endif(Q_WS_MAC)
 
diff --git a/src/collection/sqlcollection/CMakeLists.txt b/src/collection/sqlcollection/CMakeLists.txt
index 9ca3f58..fb750d5 100644
--- a/src/collection/sqlcollection/CMakeLists.txt
+++ b/src/collection/sqlcollection/CMakeLists.txt
@@ -56,29 +56,17 @@ SET(CMAKE_SHARED_LINKER_FLAGS ${CMAKE_SHARED_LINKER_FLAGS_NOFATALWARN} )
 STRING(REPLACE "-Wl,--fatal-warnings" "" CMAKE_MODULE_LINKER_FLAGS_NOFATALWARN ${CMAKE_MODULE_LINKER_FLAGS})
 SET(CMAKE_MODULE_LINKER_FLAGS ${CMAKE_MODULE_LINKER_FLAGS_NOFATALWARN} )
 
+STRING(REPLACE "-Wl,--no-undefined" "" CMAKE_MODULE_LINKER_FLAGS_ALLOW_UNDEFINED ${CMAKE_MODULE_LINKER_FLAGS})
+SET(CMAKE_MODULE_LINKER_FLAGS ${CMAKE_MODULE_LINKER_FLAGS_ALLOW_UNDEFINED} )
+
 target_link_libraries(amarok_collection-sqlcollection
     amaroklib
     amarokpud
     ${KDE4_KDEUI_LIBS}
     ${KDE4_KIO_LIBS}
     ${KDE4_THREADWEAVER_LIBRARIES}
-    ${MYSQL_EMBEDDED_LIBRARIES}
-    #this line may be needed on Debian or other systems where mysql is compiled
-    #in the prescence of yassl. However it seems to cause problems for other people,
-    #so its been disabled.
-    #See http://bugs.mysql.com/bug.php?id=21489
-    ${MYSQL_LIBRARIES}
-    ${CMAKE_DL_LIBS}
-    ${ZLIB_LIBRARIES}
 )
 
-if(NOT WIN32)
-    target_link_libraries(amarok_collection-sqlcollection crypto ssl)
-    if(NOT APPLE)
-        target_link_libraries(amarok_collection-sqlcollection crypt pthread)
-    endif(NOT APPLE)
-endif(NOT WIN32)
-
 if(APPLE)
 	SET_TARGET_PROPERTIES(amarok_collection-sqlcollection PROPERTIES LINK_FLAGS "-undefined dynamic_lookup")
 endif(APPLE)
-- 
tg: (d1aa295..) debian/mysqle_link_to_amarok_executable_feature (depends on: upstream)
